module/zfs/vdev_raidz.c

```diff
@@ -2745,6 +2846,156 @@ vdev_raidz_worst_error(raidz_row_t *rr)
+{
+	uint64_t m;
+
+	if (n % 2 == 0)
+		m = (data[(n>>1) - 1] + data[n>>1]) >> 1;
+	else
+		m = data[((n + 1) >> 1) - 1];
+
+	return (m);
+}
+
+/*
+ * Calculate the outlier fence from a set of n latency values
+ *
+ * fence = Q3 + 2 x (Q3 - Q1)
+ */
+static uint64_t
+latency_quartiles_fence(const uint64_t *data, size_t n)
+{
+	uint64_t q1 = latency_median_value(&data[0], n >> 1);
+	uint64_t q3 = latency_median_value(&data[(n + 1) >> 1], n >> 1);
+
+	/*
+	 * To avoid detecting false positive outliers when N is small and
+	 * and the latencies values are very close, make sure the fence
+	 * is at least 25% larger than Q1.
+	 */
+	uint64_t iqr = MAX(q3 - q1, q1 >> 3);
+
+	return (q3 + (iqr << 1));
+}
+
+#define	LAT_SAMPLES_STACK	64
+#define	LAT_SAMPLES_MIN		5
+#define	LAT_OUTLIER_LIMIT	50
+
+static int
+latency_compare(const void *arg1, const void *arg2)
+{
+	const uint64_t *l1 = (uint64_t *)arg1;
+	const uint64_t *l2 = (uint64_t *)arg2;
+
+	return (TREE_CMP(*l1, *l2));
+}
+
+/*
+ * Check for any latency outlier from latest set of child reads.
+ *
+ * Uses a Tukey's fence, with K = 2, for detecting extreme outliers. This
+ * rule defines extreme outliers as data points outside the fence of the
+ * third quartile plus two times the Interquartile Range (IQR). This range
+ * is the distance between the first and third quartile.
+ */
+static void
+vdev_child_slow_outlier(zio_t *zio)
+{
+	vdev_t *vd = zio->io_vd;
+	if (vdev_read_sit_out_secs == 0 || vd->vdev_children < LAT_SAMPLES_MIN)
+		return;
+
+	hrtime_t now = gethrtime();
+	uint64_t last = atomic_load_64(&vd->vdev_last_latency_check);
+
+	if ((now - last) < MSEC2NSEC(raid_outlier_check_interval_ms) ||
+	    atomic_cas_64(&vd->vdev_last_latency_check, last, now) != last) {
+		return;
+	}
+
+	int samples = vd->vdev_children;
+	uint64_t data[LAT_SAMPLES_STACK];
+	uint64_t *lat_data;
+
+	if (samples > LAT_SAMPLES_STACK)
+		lat_data = kmem_alloc(sizeof (uint64_t) * samples, KM_SLEEP);
+	else
+		lat_data = &data[0];
+
+	uint64_t max = 0;
+	vdev_t *svd = NULL; /* suspect vdev */
+	for (int c = 0; c < samples; c++) {
+		vdev_t *cvd = vd->vdev_child[c];
+
+		if (cvd->vdev_read_sit_out_expire != 0) {
+			if (cvd->vdev_read_sit_out_expire < gethrestime_sec()) {
+				/*
+				 * Done with our sit out, wait for new outlier
+				 * to emerge.
+				 */
+				cvd->vdev_read_sit_out_expire = 0;
+			} else {
+				atomic_store_64(&cvd->vdev_ewma_latency, 0);
+				/* Only one sit out disk at a time for now */
+				goto out;
+			}
+		}
+
+		lat_data[c] = atomic_load_64(&cvd->vdev_ewma_latency);
+
+		/* wait until all disks have been read from */
+		if (lat_data[c] == 0)
+			goto out;
+
+		/* keep track of the vdev with largest value */
+		if (lat_data[c] > max) {
+			max = lat_data[c];
+			svd = cvd;
+		}
+	}
+
+	qsort((void *)lat_data, samples, sizeof (uint64_t), latency_compare);
+	uint64_t fence = latency_quartiles_fence(lat_data, samples);
+	if (lat_data[samples - 1] > fence) {
```

Comment:
We should be able to sit out multiple drives.  What I think we want is to identify the worst outliers here (up to the parity level) and set `->vdev_read_sit_out_expire` on them.  Then in `vdev_*_io_start_read*` function we make the last minute decision if we need to issue an IO to this child.  It's only when issuing the IO can we fully take in to account failed drives, known missing data, scrub/resilver IOs, IO errors, etc.  Setting `->vdev_read_sit_out_expire` should be considered additional advisory information and we can (and should) read this device if we need too.